draft-ietf-mls-protocol.md

```diff
@@ -1444,110 +1452,169 @@ To authenticate a message from a particular member, signatures are required.
+application_secret.)
 
-Updating the Application secret and deriving the associated AEAD key and nonce can
-be summarized as the following Application key schedule where
-each member's Application secret chain looks as follows after the initial
-derivation:
+~~~~
+astree_node_[IndexOf(root)]_secret = application_secret
+~~~~
 
-~~~~~
-           application_secret_[sender]_[N-1]
-                     |
-                     +--> HKDF-Expand-Label(.,"nonce", "", nonce_length)
-                     |    = write_nonce_[sender]_[N-1]
-                     |
-                     +--> HKDF-Expand-Label(.,"key", "", key_length)
-                     |    = write_key_[sender]_[N-1]
-                     V
-           HKDF-Expand-Label(., "app sender", [sender], Hash.length)
-                     |
-                     V
-           application_secret_[sender]_[N]
-~~~~~
+The secret of any other node in the tree is derived from its parent's secret
+using a call to Derive-Secret. The context for the call is the (hash of the)
+Group state of the epoch and the index of the node whose secret is being
+derived.
 
-The Application context provided together with the previous Application secret
-is used to bind the Application messages with the next key and add some freshness.
+~~~~
+struct {
+  opaque gshash<0..2^32-1> = Hash(GroupState_[n]);
+  uint16 node_index;
+} ASTreeContext
+~~~~
 
-[[OPEN ISSUE: The HKDF context field is left empty for now.
-A proper security study is needed to make sure that we do not need
-more information in the context to achieve the security goals.]]
+If V is a node in the ASTree then ASTreeContext[V] denotes an
+ASTreeContext variable with:
 
-[[ OPEN ISSUE: At the moment there is no contributivity of Application secrets
-chained from the initial one to the next generation of Epoch secret. While this
-seems safe because cryptographic operations using the application secrets can't
-affect the group init_secret, it remains to be proven correct. ]]
+* gshash = Hash(GroupState_[n])
 
-### Updating the Application Secret
+* node_index = IndexOf(V)
 
-The following rules apply to an Application Secret:
+Moreover, the secrets of the children of V are defined to be:
 
-- Senders MUST only use the Application Secret once and monotonically
-  increment the generation of their secret. This is important to provide
-  Forward Secrecy at the level of Application messages. An attacker getting
-  hold of a member specific Application Secret at generation [N+1] will not be
-  able to derive the member's Application Secret [N] nor the associated
-  AEAD key and nonce.
+~~~~
+astree_node_[IndexOf(V)]_secret
+        |
+        |
+        +--> Derive-Secret(., "astree-secret", ASTreeContext[V.leftChild])
+        |    = astree_node_[IndexOf(V.leftChild)]_secret
+        |
+        V
+Derive-Secret(., "astree-secret", ASTreeContext[V.rightChild])
+= astree_node_[IndexOf(V.rightChild)]_secret
+~~~~
```

Comment:
Here was my thought process which lead to the schedule in the PR.

A property of left/right tree schedule i'm not a fan of: if nodes v and u end up with the same key then their children (and grandchildren, etc.) will have the same key too.

So add IndexOf(V.left_child) as context to Derive-Secret. However this still means if, in 2 different epochs a node in the AS Tree ends has the same key then so will its children. Thats why I added the Hash(GroupState_[n]).

To be clear, its not like I have a concrete attack. Its more of a "best practice" & "defense in depth" situation. IMO different pairs of keys having the same relation to each other make me nervous.

I also hoped that since Hash(GroupState_[n]) only needs to computed once, is a short string (e.g. fits into even lower cache levels) and can be blindly plugged in to all Derive-Secret calls the price for this approach wasn't too great.

So for now I've not made any changes here. However, I dont fell all *that* strongly about it. So if you do, or there's push back from others as well then feel free to change it.

Reply:
I can see why it's appealing to have each derivation entail a different relation.  Maybe we can accommodate that a bit more elegantly by moving the GroupContext hash into the definition of Derive-Secret (since that's all that is used for context anyway), and defining a Derive-Application-Secret function that gives us the knobs we need to vary things as we go down the tree and the ratchets.  Here's a sketch that should be copy-paste-able into the text here:

```
HKDF-Expand-Label(Secret, Label, Context, Length) =
    HKDF-Expand(Secret, HkdfLabel, Length)

Where HkdfLabel is specified as:

struct {
    uint16 length = Length;
    opaque label<7..255> = "mls10 " + Label;
    opaque context<0..2^32-1> = Context;
} HkdfLabel;

Derive-Secret(Secret, Label) =
    HKDF-Expand-Label(Secret, Label, Hash(GroupContext_[n]), Hash.length)

struct {
  uint32 node;
  uint32 generation;
  opaque role<0..255>;
} ApplicationContext;

Derive-Application-Secret(secret, node, generation, role) =
  DeriveSecret(secret, ApplicationContext(node, generation, role) 

astree_node_[IndexOf(V)]_secret
        |
        |
        +--> Derive-Application-Secret(., V.left_child, 0, "tree")
        |    = astree_node_[IndexOf(V.left_child)]_secret
        |
        +--> Derive-Application-Secret(., V.right_child, 0, "tree")
             = astree_node_[IndexOf(V.right_child)]_secret

application_[i]_[j]_secret
      |
      +--> Derive-Application-Secret(., 2*i, j, "key")
      |    = application_[i]_[j+1]_nonce
      |
      +--> Derive-Application-Secret(., 2*i, j, "nonce")
      |    = application_[i]_[j+1]_key
      |
      V
HKDF-Application-Secret(., 2*i, j, "secret")
= application_[i]_[j+1]_secret
```

Reply:
Here is an proposal for this which is an hybrid of your two suggestions:
```
HKDF-Expand-Label(Secret, Label, Context, Length) =
    HKDF-Expand(Secret, HkdfLabel, Length)

Where HkdfLabel is specified as:

struct {
    uint16 length = Length;
    opaque label<7..255> = "mls10 " + Label;
    opaque context<0..2^32-1> = Context;
} HkdfLabel;

Derive-Secret(Secret, Label) =
    HKDF-Expand-Label(Secret, Label, Hash(GroupContext_[n]), Hash.length)

struct {
  uint32 node;
  uint32 generation;
  opaque role<0..255>;
} ApplicationContext;

Derive-Application-Secret(secret, node, generation, role) =
  DeriveSecret(secret, ApplicationContext(node, generation, role) 

astree_node_[IndexOf(V)]_secret
        |
        |
        +--> Derive-Application-Secret(., V.left_child, 0, "tree")
        |    = astree_node_[IndexOf(V.left_child)]_secret
        |
        +--> Derive-Application-Secret(., V.right_child, 0, "tree")
             = astree_node_[IndexOf(V.right_child)]_secret


application_[i]_[j]_secret
      |
      +--> HKDF-Expand-Label(.,"nonce", "", nonce_length)
      |    = write_nonce_[i]_[j]
      |
     +--> HKDF-Expand-Label(.,"key", "", key_length)
     |    = write_key_[i]_[j] 
    V
Derive-Application-Secret(., 2*i, 0, "sender")
= application_[i]_[j+1]_secret

```
This is slightly different in the sense that we have basically don't change our current way of deriving keys and nonces from an application secret.